### PR TITLE
feat(mdns): emit `ToSwarm::NewExternalAddrOfPeer`

### LINKED ISSUE #5104: mDNS: emit `ToSwarm::NewExternalAddrOfPeer`
### Description

With https://github.com/libp2p/rust-libp2p/pull/4371 merged, we have the foundation of https://github.com/libp2p/rust-libp2p/issues/4302 implemented.

`libp2p-mdns` needs to be extended to emit this new event whenever it discovers a new address via mDNS.

### Motivation

Other behaviours should be able to learn the addresses discovered by mDNS.

### Current Implementation

Only mDNS knows about its discovered addresses.

### Are you planning to do it yourself in a pull request ?

No

### PR DESCRIPTION
**Pull request by justcode740**
## Description
Resolves #5104 

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/mdns/src/behaviour.rs [hunk 1]
```diff
@@ -347,8 +347,10 @@ where
         }
 
         if !discovered.is_empty() {
-            let event = Event::Discovered(discovered);
-            return Poll::Ready(ToSwarm::GenerateEvent(event));
+            return Poll::Ready(ToSwarm::NewExternalAddrOfPeer {
```
Comment (retrohacker) [inline protocols/mdns/src/behaviour.rs:350]: [question]

Sorry if this is a silly question, I'm new to both rust and the libp2p code base.

1) This seems like a breaking change for any code base that currently relies on `Event::Discovered`

Is my understanding of rust breaking changes right here?

Would it be worth preserving `Event::Discovered` here for backwards compatibility?

2) This only emits the first discovered peer

Reading lines 333-345, it looks like the Poll event will keep going until it drains query_response_receiver, discovering multiple peers?

This seems like we would want to return a `ToSwarm::NewExternalAddrOfPeer` for every peer we discover on line 345.

Should this move up to line 345 and replace the discovered vec?

3) I acknowledge that, as written, point 2 might be at odds with point 1 :sweat_smile:.

To have our cake, and eat it too, could 333-347 be moved out of `fn poll` and instead feed two receivers, one that sends `Event::Discovered` after all `ToSwarm::NewExternalAddrOfPeer`s have been flushed and the other that sends each `ToSwarm::NewExternalAddrOfPeer`?

Comment (jxs) [reply protocols/mdns/src/behaviour.rs:350]: yeah exactly, thanks @retrohacker!
You can see the [upnp example](https://github.com/libp2p/rust-libp2p/blob/master/protocols/upnp/src/behaviour.rs#L422-L435) on how to return both

**Comment by dariusc93** [comment]
Any updates to this @justcode740?

**Comment by jxs** [comment]
superseeded by #5753

**Pull request closed without merging**